### ISSUE TITLE
autoencoder with state-of-the-art optimizer

### ISSUE THREAD

**Issue by zealseeker** [Shark-ML/Shark#181]
In the [tutorial of autoencoder](http://image.diku.dk/shark/sphinx_pages/build/html/rest_sources/tutorials/algorithms/autoencoders.html) the optimizer is `IRpropPlusFull` and I find that it's an "old" algorithm ([Improving the Rprop Learning Algorithm](http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.17.1332))
In tensorflow or other "hot" deep learning toolkits, the optimizer includes "SGD, Adam, AdamDelta, RMSprop, Adamax, etc.". I don't know which is better but, they are widely used in current deep learning modeling.
So I wonder: 1) Will SHARK support the state-of-the-art optimizer in future? 2) Is IRpropPlus similar with the one of these methods? 3) Is it necessary to use the state-of-the-art optimizer in autoencoder, from the aspect of computing performance and "accuracy" (The accuracy is hard to define since it's an unsupervised method)

I tried understanding the these algorithms... but nothing works. 
I used `IRpropPlusFull` to train 60k instances with 166 features for 300 steps. It takes about one hour.

**Comment by Ulfgard** [comment]
Hi,

1) We plan to support some of the newer versions of SGD (but note that we support basic SGD already by using the SteepestDescent optimizer together with the NoisyErrorFunction). Right now, Shark is not particularly optimized wrt SGD, especially with small batch sizes. So we want to see whether we can optimize this first.

3) In general, SGD comes at the price that the conditioning of the problem becomes a major problem and you can see this already in pure linear problems. E.g. if you want to do a statistical tests on logistic regression, don't use SGD. Even SAG can become problematic. This problem is largely overlooked in large scale machine learning, because:
  - We never train until convergence
  - We have other sources of noise which prevent linear convergence rates(Dropout)
  - We can change the model to fit the optimizer
So SGD is fine if you train truely large-scale, use linear rectified units and do not care about precision in terms of the loss function you optimize. Otherwise, the "old" algorithms will quite often win.

3) RProp performs on Autoencoders often similar to BFGS and LBFGS, and like them it is not an SGD method. It comes at the advantage that it is invariant to monotonous transformations of the error function. This means especially that it does not suffer from vanishing or exploding gradients, which makes it quite robust. Personally i often use LBFGS, but this is a matter of taste.


4) Your performance looks not very good, i think. Please make sure the following is given:
- OpenMP is activated
- You use some advanced linear algebra package like ATLAS or OpenBLAS. Unfortunately as systems put stuff in different weird places, it might be that Shark does not find it. In that case I would try to fix it for you.
- BatchSize is large. We use a very conservative batch size of 256 by default (subject to change). We often perform a lot better with a batchsize of 4096 or so.  Typically you can just add the batchSize parameter when you load the dataset.

**Comment by zealseeker** [comment]
Thank you very much for the detailed explanations.
I'll try to use SGD and compare them with IRprop+. Now I have understood the algorithm of IRprop+ from the paper. But [(L)BFGS](https://en.wikipedia.org/wiki/Limited-memory_BFGS) seems too difficult for me ...

There's another small question that I want to consult about:
According to IRprop+, eta+, eta-, delta_max, and delta_min are predefined hyper-parameters. But in the reference of  [IRpropPlus](http://image.diku.dk/shark/doxygen_pages/html/classshark_1_1_i_rprop_plus_full.html#a6e96d428dd477bfd94245d44afc3ab4b), there is no such argument.

As for the performance, our cluster do not support BLAS and we don't know how to install it as a limited user ("using `root` may affect the environment and disturb others' work", said by the administrator). Currently the performance is acceptable since we can submit several jobs parallel.

**Comment by Ulfgard** [comment]
This is fixed in DeepLearning branch, see the Autoencoder tutorial there. Should look very close to Keras or Tensorflow.
